---
title: "RNA-seq exh samples"
author: "German Novakovskiy"
date: "August 27, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#cleaning the environment
rm(list=ls())

#loading required packages
library(limma) 
library(edgeR)
library(ggplot2)
library(dplyr)
library(RColorBrewer)
library(tibble)
library(pheatmap)
library(sva)
library(fgsea)
library(stats)
library(gplots)
library(RSvgDevice)
library(xlsx)
library(stringr)
library(DESeq)
library(statmod)
library(pcaMethods)
library(fastICA)
library(RSvgDevice)
library(cairoDevice)
library(ggplot2)
```


## RNA-seq/Caroline data

```{r}
#loading files
files <- list.files("Caro RNAseq/")
x <- readDGE(files, path = "Caro RNAseq/", columns=c(1,2), header=F)

#removing meta tags
x$counts <- x$counts[1:(nrow(x$counts)-2),]

#loading metadata
meta_data <- read.xlsx("RNAseq sample info.xlsx", sheetName = "CAR-Treg RNAseq sample info")
meta_data <- as.data.frame(meta_data)

#fixing file name so it's the same as in data frame
meta_data$FileName <- sapply(as.character(meta_data$FileName), 
                             function(x){ 
                               paste(unlist(strsplit(x, split="[.]"))[1:3], collapse=".") 
                               })
rownames(meta_data) <- meta_data$FileName
meta_data <- meta_data[rownames(x$samples),]

#just for simplicity
#and For plotting MDS plot
donor <- as.factor(meta_data$Donor)
batch <- as.factor(meta_data$Batch)
cell <- as.factor(meta_data$Cell)
construct <- as.factor(meta_data$Construct)

x$samples$Donor <- donor
x$samples$Batch <- batch
x$samples$Cell <- cell
x$samples$Construct <- construct

#Filtering lowly expressed genes
#using construct is a main group
keep.exprs <- filterByExpr(x, group=construct)
x <- x[keep.exprs,, keep.lib.sizes=FALSE] #12792    18

#Normalizing for the library size
x <- calcNormFactors(x, method = "TMM")

#Building MDS plot
lcpm <- cpm(x, log=TRUE)
col.donor <- x$samples$Donor
levels(col.donor) <-  brewer.pal(nlevels(col.donor), "Set1")
col.donor <- as.character(col.donor)
plotMDS(lcpm, cex=0.75, labels = donor, col = col.donor, dim.plot = c(1,2)) #first and second dimension
#plotMDS(lcpm, cex=0.75, labels = construct, col = col.donor, dim.plot = c(3,4)) #third and fourth
```
```{r}
col.construct <- x$samples$Construct
levels(col.construct) <-  brewer.pal(nlevels(col.construct), "Set1")
col.construct <- as.character(col.construct)
plotMDS(lcpm, cex=0.75, labels = construct, col = col.construct, dim.plot = c(1,2)) #first and second dimension
```
```{r}
col.batch <- x$samples$Batch
levels(col.batch) <-  brewer.pal(nlevels(col.batch), "Set1")
col.batch <- as.character(col.batch)
plotMDS(lcpm, cex=0.75, labels = batch, col = col.batch, dim.plot = c(1,2)) #first and second dimension
```

```{r}
col.cell <- x$samples$Cell
levels(col.cell) <-  brewer.pal(nlevels(col.cell), "Set1")
col.cell <- as.character(col.cell)
plotMDS(lcpm, cex=0.75, labels = cell, col = col.cell, dim.plot = c(1,2)) #first and second dimension
```
```{r}
#all together
devSVG(file = "/home/german/Caroline_RNA_seq/Caro RNAseq/MDS_plot_all_labels_batch.svg")
all_labels <- str_c(donor,"/",construct, "/", batch, "/", cell)
Cairo(filename="/home/german/Caroline_RNA_seq/Caro RNAseq/MDS_plot_all_labels_batch.svg", 
       surface="svg", 
       width=12, 
       height=10, 
       pointsize=12)
plotMDS(lcpm, cex=0.75, labels = all_labels, col = col.donor, dim.plot = c(1,2)) #first and second dimension
dev.off()
```
```{r}
mds <- plotMDS(lcpm, cex=0.75, labels = all_labels, col = col.donor, dim.plot = c(1,2)) 
toplot <- data.frame(Dim1 = mds$x, Dim2 = mds$y, Cell = meta_data$Cell, Construct = meta_data$Construct)

devSVG(file = "/home/german/Caroline_RNA_seq/Caro RNAseq/MDS_plot_all_labels_batch_GGPLOT.svg")
Cairo(filename="/home/german/Caroline_RNA_seq/Caro RNAseq/MDS_plot_all_labels_batch_GGPLOT.svg", 
       surface="svg", 
       width=12, 
       height=10, 
       pointsize=12)
ggplot(toplot, aes(Dim1, Dim2, colour = Construct)) + 
  geom_point(aes(shape = Cell), size=4) +
  theme_bw() +
  xlab("Leading logFC dim 1") + 
  ylab("Leading logFC dim 2")
#
```


```{r}
meta_data$Labels <- all_labels

#just to visualize batch correction
lcpm.no.batches <- removeBatchEffect(lcpm, batch)
lcpm.no.donors <- removeBatchEffect(lcpm, donor)
lcpm.no.cell <- removeBatchEffect(lcpm, cell)

#
devSVG(file = "/home/german/Caroline_RNA_seq/Caro RNAseq/MDS_plot_all_labels_no_batch.svg")
Cairo(filename="/home/german/Caroline_RNA_seq/Caro RNAseq/MDS_plot_all_labels_no_batch.svg", 
       surface="svg", 
       width=12, 
       height=10, 
       pointsize=12)
plotMDS(lcpm.no.batches, cex=0.75, labels = all_labels, col = col.donor, dim.plot = c(1,2)) #first and second dimension
dev.off()
```
```{r}
mds <- plotMDS(lcpm.no.batches, cex=0.75, labels = all_labels, col = col.donor, dim.plot = c(1,2)) 
toplot <- data.frame(Dim1 = mds$x, Dim2 = mds$y, Cell = meta_data$Cell, Construct = meta_data$Construct)

devSVG(file = "/home/german/Caroline_RNA_seq/Caro RNAseq/MDS_plot_all_labels_no_batch_GGPLOT.svg")
Cairo(filename="/home/german/Caroline_RNA_seq/Caro RNAseq/MDS_plot_all_labels_no_batch_GGPLOT.svg", 
       surface="svg", 
       width=12, 
       height=10, 
       pointsize=12)
ggplot(toplot, aes(Dim1, Dim2, colour = Construct)) + 
  geom_point(aes(shape = Cell), size=4) +
  theme_bw() +
  xlab("Leading logFC dim 1") + 
  ylab("Leading logFC dim 2")
#dev.off()
```

```{r}
#PCA plot
project.pca <- prcomp(t(lcpm), center = TRUE,scale. = TRUE)
project.pca.proportionvariances <- ((project.pca$sdev^2) / (sum(project.pca$sdev^2)))*100

plot(project.pca$x, type="n", main="Principal components analysis bi-plot", xlab=paste("PC1, ", round(project.pca.proportionvariances[1], 2), "%"), ylab=paste("PC2, ", round(project.pca.proportionvariances[2], 2), "%"))
points(project.pca$x, col=c("red","blue","green")[donor], pch=16, cex=1)
text(project.pca$x, col=c("red","blue","green")[donor], labels = all_labels, pch=16, cex=0.75, pos = 3)
```
```{r}
project.pca <- prcomp(t(lcpm.no.batches), center = TRUE,scale. = TRUE)
project.pca.proportionvariances <- ((project.pca$sdev^2) / (sum(project.pca$sdev^2)))*100

plot(project.pca$x, type="n", main="Principal components analysis bi-plot", xlab=paste("PC1, ", round(project.pca.proportionvariances[1], 2), "%"), ylab=paste("PC2, ", round(project.pca.proportionvariances[2], 2), "%"))
points(project.pca$x, col=c("red","blue","green")[donor], pch=16, cex=1)
text(project.pca$x, col=c("red","blue","green")[donor], labels = all_labels, pch=16, cex=0.75, pos = 3)
```

```{r}
#get the most variable genes
lib.size <- estimateSizeFactorsForMatrix(x$counts)
ed <- t(t(x$counts)/lib.size)

#Calculate estimates of variance, coefficient of variation
means <- rowMeans(ed)
vars <- apply(ed,1,var)
cv2 <- vars/means^2

#Now fit a regression line based on the controls:
minMeanForFit <- unname( quantile( means[ which( cv2 > .3 ) ], .95 ) )
useForFit <- means >= minMeanForFit # & spikeins
fit <- glmgam.fit( cbind( a0 = 1, a1tilde = 1/means[useForFit] ),cv2[useForFit] )
a0 <- unname( fit$coefficients["a0"] )
a1 <- unname( fit$coefficients["a1tilde"])
df <- ncol(ed) - 1

#Rank genes by the significance of deviation from the fit
afit <- a1/means+a0
varFitRatio <- vars/(afit*means^2)
varorder <- order(varFitRatio,decreasing=T)
oed <- ed[varorder,]

#We can also evaluate statistical significance of the deviation
pval <- pchisq(varFitRatio*df,df=df,lower.tail=F)
adj.pval <- p.adjust(pval,"fdr")
sigVariedGenes <- adj.pval<1e-3

#heatmap construction
#lcpm.no.batches <- removeBatchEffect(lcpm, batch)

#colnames(lcpm.no.batches) <- all_labels #uncomment if want to do without batches
colnames(lcpm) <- all_labels

#varGenes <- lcpm.no.batches[rownames(oed[1:50,]),] #uncomment if want to do without batches
varGenes <- lcpm[rownames(oed[1:50,]),]
#scaling the data (without it in pheatmap specify scale = "row")
varGenes <- t(scale(t(varGenes)))

my_sample_col <- data.frame(cell = cell, donor = donor, construct = construct)
#row.names(my_sample_col) <- colnames(lcpm.no.batches) #uncomment if want to do without batches
row.names(my_sample_col) <- colnames(lcpm)

devSVG(file = "/home/german/Caroline_RNA_seq/Caro RNAseq/heatmap_plot_no_batch.svg")
pheatmap(varGenes, cluster_rows = T, cluster_cols = T, 
         color=bluered(21), scale = "none", clustering_method = "ward.D2", 
         clustering_distance_cols = "euclidean", show_colnames = T, show_rownames = T, 
         annotation_col = my_sample_col,
         main = "50 most variable genes")
dev.off()
```

```{r}
#Help functions
###############

#function for DGE for specified question (groups)
limma_DGE_group_analysis <- function(expr_mat, meta_data, group1, group2){
  #Introducing the corresponding column to meta data 
  md <- meta_data
  md <- md %>% filter(Labels %in% c(group1, group2))
  md$Analysis <- integer(nrow(md))
  md$Analysis[which(md$Labels %in% group1)] <- 1
  md$Analysis[which(md$Labels %in% group2)] <- 2
  md$Analysis <- as.factor(md$Analysis)  
  
  #specifying design matrix; regress out Donor (since it's nested in batch, it will take into account batch effects); 
  #no intercept (0) so easier to look at differences between groups
  #designMatrix <- model.matrix(~0 + Analysis + Batch, md)
  designMatrix <- model.matrix(~0 + Analysis, md) #differences are not that big to correct for
  #contrast matrix to specify what type of differences we are looking for 
  contrastMatrix <- makeContrasts(
    A1 = Analysis1 - Analysis2,
    levels = colnames(designMatrix))
  
  #subsetting the corresponding samples
  expr_mat <- expr_mat[,md$FileName]
  
  #voom -  estimates the mean-variance relationship and uses this to compute appropriate observation-level weights
  v <- voom(expr_mat, designMatrix, plot=FALSE)
  #Fit linear model for each gene
  vfit <- lmFit(v, designMatrix)
  #fit the contrast using the original fitted model
  vfit <- contrasts.fit(vfit, contrasts=contrastMatrix)
  #apply eBayes() for moderated statistics
  efit <- eBayes(vfit)
  
  #0.05 adjusted p-value
  cutoff <- 5e-02 
  #Multiple Test Correction method by default is BH (equivalent to FDR)
  dt <- decideTests(efit, p.value = cutoff, lfc = 1)
  dt <- as.data.frame(dt)
  #table with all genes and their DGE stats
  DEgenes <- topTable(efit, number = Inf)
  
  #calculating log-cpm
  lcpm <- cpm(expr_mat, log=TRUE)
  
  #for input
  res <- list()
  res[["DecisionTable"]] <- dt
  res[["DEgenes"]] <- DEgenes
  #res[["LogCPM"]] <- v$E
  res[["LogCPM"]] <- lcpm
  res[["eFit"]] <- efit
  
  return(res)
}
###############

###############
#function for gene set enrichment analysis (GSEA); TF - do for TFs or no?
#terms is required - it's a gmt file with TF targets or Pathway terms
enrichmentAnalysis <- function(DEgenes, TF = F, terms = NULL, minsize){
  
  gseaInputGeneScores <- DEgenes %>%
    dplyr::select("logFC", "adj.P.Val", "P.Value", "gene")
  
  gseaInputGeneScores$Sign <- sapply(gseaInputGeneScores$logFC, sign)
  #gseaInputGeneScores$LogPvalue <- sapply(gseaInputGeneScores$adj.P.Val, function(x){ -log10(x) }) #you can do adjusted/ no big difference actually
  gseaInputGeneScores$LogPvalue <- sapply(gseaInputGeneScores$P.Val, function(x){ -log10(x) })
  gseaInputGeneScores$Score <- gseaInputGeneScores$Sign * gseaInputGeneScores$LogPvalue 
  
  #create ranked list of genes
  #gseaInputGeneScores <- DEgenes %>% 
    #mutate(absolute_logFC = abs(logFC)) %>% 
    #dplyr::select(gene, t) %>% 
    #na.omit() %>% 
    #as.data.frame()
  
  genes <- gseaInputGeneScores$gene
  gseaInputGeneScores <- gseaInputGeneScores$Score
  names(gseaInputGeneScores) <- genes
  gseaInputGeneScores <- sort(gseaInputGeneScores, decreasing = T)
  #write(genes, file="~/Desktop/genes.txt")
  
  #perform GSEA 
  fgseaRes <- fgsea(terms, gseaInputGeneScores, minSize=minsize, maxSize=300, nperm=10000)
  
  #filter only statistically significant terms and look at direction (NES > 0 means enrichment at the top of ranked list)
  sign_pos <- fgseaRes %>% filter(padj < 0.05 & NES > 0)
  sign_neg <- fgseaRes %>% filter(padj < 0.05 & NES < 0)
  
  #sort by NES (normalized enrichment score)
  sign_pos <- sign_pos %>% arrange(desc(NES))
  sign_neg <- sign_neg %>% arrange(NES)
  
  #saving resulting data frames to list
  res <- list()
  res[["Genes"]] <- genes
  res[["Enrichment"]] <- fgseaRes
  res[["Scores"]] <- gseaInputGeneScores
  res[["Sign_Pos"]] <- sign_pos
  res[["Sign_Neg"]] <- sign_neg
  return(res)
}
###############

#######################################################
#MAIN DGE FUNCTION
#######################################################
performDGEanalysis <- function(expr_mat, meta_data, group1, group2, termGO = NULL,
                               termTF = NULL, termH = NULL){
  #perform DGE analysis with limma
  test <- limma_DGE_group_analysis(expr_mat, meta_data, group1, group2)
  print("Done with limma analysis")
  #get decision table
  dt <- test[["DecisionTable"]]
  #get data frame with gene expression stats
  DEgenes <- test[["DEgenes"]]
  #get log cpm matrix
  log_cpm <- test[["LogCPM"]]
  eFit <- test[["eFit"]]
  
  #arrange genes by logFC
  DEgenes <- DEgenes %>% rownames_to_column("gene") %>% arrange(desc(logFC))
  
  #df of significantly up regulated genes
  upRegulatedGenes <- DEgenes %>% filter(logFC > 1 & adj.P.Val < 0.05)
  #df of significantly down regulated genes
  downRegulatedGenes <- DEgenes %>% filter(logFC < -1 & adj.P.Val < 0.05)
  
  enrichedGoPathways <- enrichmentAnalysis(DEgenes, TF = F, terms = termGO, minsize = 15)
  print("Done with Pathways enrichment")
  sign_pos_pathways <- enrichedGoPathways[["Sign_Pos"]]
  sign_neg_pathways <- enrichedGoPathways[["Sign_Neg"]]
  
  enrichedTFs <- enrichmentAnalysis(DEgenes, TF = T, terms = termTF, minsize = 5)
  print("Done with TFs enrichment")
  sign_pos_TF <- enrichedTFs[["Sign_Pos"]]
  sign_neg_TF <- enrichedTFs[["Sign_Neg"]]
  TFfgsea <- enrichedTFs[["Enrichment"]]
  TFscores <- enrichedTFs[["Scores"]]
  
  enrichedHPathways <- enrichmentAnalysis(DEgenes, TF = F, terms = termH, minsize = 15)
  print("Done with Hallmark Pathways enrichment")
  sign_pos_pathwaysH <- enrichedHPathways[["Sign_Pos"]]
  sign_neg_pathwaysH <- enrichedHPathways[["Sign_Neg"]]
  
  #genes for GOrilla analysis (or any other GO enrichment tool)
  genes <- enrichedGoPathways[["Genes"]]
  
  #saving results
  res <- list()
  res[["DEgenesTable"]] <- DEgenes
  res[["upRegulatedGenes"]] <- upRegulatedGenes
  res[["downRegulatedGenes"]] <- downRegulatedGenes
  res[["Sign_Pos_Pathway"]] <- sign_pos_pathways
  res[["Sign_Neg_Pathway"]] <- sign_neg_pathways
  res[["Sign_HALLMARK_Pos_Pathway"]] <- sign_pos_pathwaysH
  res[["Sign_HALLMARK_Neg_Pathway"]] <- sign_neg_pathwaysH
  res[["Sign_Pos_TF"]] <- sign_pos_TF
  res[["Sign_Neg_TF"]] <- sign_neg_TF
  res[["TF_enrichment"]] <- TFfgsea
  res[["TF_scores"]] <- TFscores
  res[["genes"]] <- genes
  res[["DecisionTable"]] <- dt
  res[["LogCPM"]] <- log_cpm
  res[["eFit"]] <- eFit
  return(res)
}
```

```{r}
#upRegulatedGenes_ut_treg_cd4
#downRegulatedGenes_ut_treg_cd4

#upRegulatedGenes_ha_treg_cd4
#downRegulatedGenes_ha_treg_cd4

#write(upRegulatedGenes_ut_treg_cd4$gene, "for_Venn_diagram/upRegulatedGenes_ut_treg_cd4.txt")
#write(upRegulatedGenes_ha_treg_cd4$gene, "for_Venn_diagram/upRegulatedGenes_ha_treg_cd4.txt")
#write(downRegulatedGenes_ut_treg_cd4$gene, "for_Venn_diagram/downRegulatedGenes_ut_treg_cd4.txt")
#write(downRegulatedGenes_ha_treg_cd4$gene, "for_Venn_diagram/downRegulatedGenes_ha_treg_cd4.txt")

#DEgenes_filtered <- DEgenes %>% filter(!gene %in% upRegulatedGenes_ut_treg_cd4$gene) %>%
#  filter(!gene %in% downRegulatedGenes_ut_treg_cd4$gene) 

#devSVG(file = "/home/german/Caroline_RNA_seq/Caro RNAseq/HA_Treg_CD4_no_UT_genes_volcanoplot.svg")
#ggplot(data = DEgenes_filtered, aes(x = logFC, y = -log(adj.P.Val), color = ((-log(adj.P.Val) > 3) & (logFC > 1 | logFC < -1))))+
#  scale_colour_manual(name = 'BH p-value < 0.05', values = setNames(c('red','black'),c(T, F)), labels = c("False", "True"))+
#  geom_point()+
#  geom_vline(xintercept=0)+
#  geom_vline(xintercept=-1)+
#  geom_vline(xintercept=1)+
#  geom_hline(yintercept=3)+
#  geom_text(aes(label = ifelse((-log(adj.P.Val) > 5) & (logFC > 6 | logFC < -6), gene, "")), vjust=-1, size = 3)+
#  #geom_text(aes(label = ifelse((-log(adj.P.Val) > 5) | (logFC > 2.5 | logFC < -2.5), gene, "")), vjust=-1, size = 3)+#(to visualize some #of the genes, note overlaps)
#  ylab("-log(p-value)")+
#  xlab("logFC")+
#  labs(title="Gene expression differences in two groups")+
#  theme_bw()+
#  theme(axis.line = element_line(colour = "black"),
#        panel.grid.major = element_blank(),
#        panel.grid.minor = element_blank(),
#        #panel.border = element_blank(),
#        panel.background = element_blank()) +
#  theme(axis.title.x=element_text(size=12),
#        axis.text.y=element_text(size=12),
#        axis.title.y=element_text(size=14),
#        axis.ticks.x=element_blank(),
#        strip.text.x = element_text(size=14),
#        strip.background = element_rect(colour="white", fill="white"),
#        legend.text=element_text(size=10),
#        legend.title=element_text(size=10))
#dev.off()
```


```{r}
#Which genes and transcriptional pathways are differentially activated in HA CAR Treg vs HA CAR CD4s?

#new thing
# Treg HA vs CD4 HA
# Treg HA vs Treg 19+UT
# Treg UT vs CD4 UT

group1A <- c("64/ha/26/Treg", "55/ha/24/Treg", "52/ha/24/Treg")
group1B <- c("64/ha/26/cd4", "55/ha/24/cd4", "52/ha/24/cd4") 

#group1A <- c("64/ha/26/Treg", "55/ha/24/Treg", "52/ha/24/Treg")
#group1B <- c("64/ut/26/Treg", "52/ut/24/Treg", "55/ut/24/Treg", "64/cd19/26/Treg", "52/cd19/24/Treg", "55/cd19/24/Treg") 

#group1A <- c("64/ut/26/Treg", "52/ut/24/Treg", "55/ut/24/Treg")
#group1B <- c("55/ut/24/cd4", "64/ut/26/cd4", "52/ut/24/cd4")

#Loading gmt files for enrichment analysis
termGO <- gmtPathways("/home/german/Nick_Treg_project/Human_GOBP_AllPathways_no_GO_iea_April_01_2019_symbol.gmt")
termTF <- gmtPathways("/home/german/Nick_Treg_project/RegNetworkStrong.gmt")
termH <- gmtPathways("/home/german/Nick_Treg_project/h.all.v6.2.symbols.gmt")

DGEres <- performDGEanalysis(x, meta_data, group1A, group1B, termGO = termGO,
                             termTF = termTF, termH = termH)

#table with all genes and their scores
DEgenes <- DGEres[["DEgenesTable"]]
#up regulated genes in group1A compared to group1B
upRegulatedGenes <- DGEres[["upRegulatedGenes"]]
#down regulated genes in group1A compared to group1B
downRegulatedGenes <- DGEres[["downRegulatedGenes"]]
#pathways positively enriched in group1A compared to group1B
sign_pos_pathways <- DGEres[["Sign_Pos_Pathway"]]
#pathways negatively enriched in group1A compared to group1B
sign_neg_pathways <- DGEres[["Sign_Neg_Pathway"]]

#pathways positively enriched in group1A compared to group1B
sign_pos_pathwaysH <- DGEres[["Sign_HALLMARK_Pos_Pathway"]]
#pathways negatively enriched in group1A compared to group1B
sign_neg_pathwaysH <- DGEres[["Sign_HALLMARK_Neg_Pathway"]]

TFgsea <- DGEres[["TF_enrichment"]]
TFscores <- DGEres[["TF_scores"]]

#Transcription factors positively enriched in group1A compared to group1B
sign_pos_TF <- DGEres[["Sign_Pos_TF"]]
#Transcription factors negatively enriched in group1A compared to group1B
sign_neg_TF <- DGEres[["Sign_Neg_TF"]]

#Decision table (potential use - Venn Diagramm; checking what genes are DE in different conditions)
dt <- DGEres[["DecisionTable"]]
#log cpm matrix for heat map construction
log_cpm <- DGEres[["LogCPM"]]
eFit <- DGEres[["eFit"]]

#saving data
write.table(upRegulatedGenes, file="/home/german/Caroline_RNA_seq/Caro RNAseq/DE_genes/upRegulated_genes_Treg_CD4.txt", sep="\t", row.names = F)
write.table(downRegulatedGenes, file="/home/german/Caroline_RNA_seq/Caro RNAseq/DE_genes/downRegulated_genes_Treg_CD4.txt", sep="\t", row.names = F)

write.table(sign_pos_TF[,-8], file="/home/german/Caroline_RNA_seq/Caro RNAseq/TF/posTF_Treg_CD4.txt", sep="\t", row.names = F)
write.table(sign_neg_TF[,-8], file="/home/german/Caroline_RNA_seq/Caro RNAseq/TF/negTF_Treg_CD4.txt", sep="\t", row.names = F)

write.table(sign_pos_pathwaysH[,-8], file="/home/german/Caroline_RNA_seq/Caro RNAseq/PathwaysHallmark/posPathways_Treg_CD4.txt", sep="\t", row.names = F)
write.table(sign_neg_pathwaysH[,-8], file="/home/german/Caroline_RNA_seq/Caro RNAseq/PathwaysHallmark/negPathways_Treg_CD4.txt", sep="\t", row.names = F)

write.table(sign_pos_pathways[,-8], file="/home/german/Caroline_RNA_seq/Caro RNAseq/Pathways/posPathways_Treg_CD4.txt", sep="\t", row.names = F)
write.table(sign_neg_pathways[,-8], file="/home/german/Caroline_RNA_seq/Caro RNAseq/Pathways/negPathways_Treg_CD4.txt", sep="\t", row.names = F)

################################################################################
#z-score calculation
#zscoreT
upRegulatedZscores <- zscoreT(upRegulatedGenes$t, unique(eFit[["df.total"]]))
names(upRegulatedZscores) <- upRegulatedGenes$gene
upRegulatedZscores <- as.data.frame(upRegulatedZscores)

downRegulatedZscores <- zscoreT(downRegulatedGenes$t, unique(eFit[["df.total"]]))
names(downRegulatedZscores) <- downRegulatedGenes$gene
downRegulatedZscores <- as.data.frame(downRegulatedZscores)

write.table(upRegulatedZscores, file="/home/german/Caroline_RNA_seq/Caro RNAseq/z_score/zscoreT/TregUT_CD4UT_upregulated.txt", 
            sep="\t", row.names = T)

write.table(downRegulatedZscores, file="/home/german/Caroline_RNA_seq/Caro RNAseq/z_score/zscoreT/TregUT_CD4UT_downregulated.txt", 
            sep="\t", row.names = T)

################################################################################
#heatmap construction
#filtering meta data to groups only
meta_data_f <- meta_data %>% filter(Labels %in% c(group1A, group1B))
#topGenesRegulated <- cleaned_log_cpm_df[common_regulated_genes,]

#correct for batch - in this case it's donor factor
#batch <- x$samples$Donor
#batch <- meta_data_f$Batch
#lcpm.no.batches <- removeBatchEffect(log_cpm, batch) #uncomment if you want to use batches

#log_cpm_f <- lcpm.no.batches #uncomment if you want to use batches
log_cpm_f <- log_cpm
colnames(log_cpm_f) <- meta_data_f$Labels
my_sample_col_an <- my_sample_col[colnames(log_cpm_f),]

upRegulatedGenes <- upRegulatedGenes$gene
downRegulatedGenes <- downRegulatedGenes$gene
common_regulated_genes <- c(upRegulatedGenes[1:25], downRegulatedGenes[1:25])
topGenesRegulated <- log_cpm_f[common_regulated_genes,]
#topGenesRegulated <- topGenesRegulated[,col.order]

#scaling the data (without it in pheatmap specify scale = "row")
topGenesRegulated <- t(scale(t(topGenesRegulated)))

write.table(topGenesRegulated, file="/home/german/Caroline_RNA_seq/Caro RNAseq/z_score/zscoreCPM/TregUT_CD4UT_zscores.txt", 
            sep="\t", row.names = T)

rownames(my_sample_col_an) <- colnames(topGenesRegulated)

#creating a heatmap with pheatmap (no clustering because we ordered samples by time point)
devSVG(file = "/home/german/Caroline_RNA_seq/Caro RNAseq/heatmap_HA_vs_UT_and_CD19.svg")
pheatmap(topGenesRegulated, cluster_rows = T, cluster_cols = T, 
         color=bluered(21), scale = "none", clustering_method = "ward.D2", 
         clustering_distance_cols = "euclidean", show_colnames = T, show_rownames = T, 
         annotation_col = my_sample_col_an,
        main = "Clustering heatmap for 50 common regulated genes")
dev.off()
```
```{r}
################################################################################
#Volcano plot construction
devSVG(file = "/home/german/Caroline_RNA_seq/Caro RNAseq/Treg_CD4_volcanoplot.svg")
ggplot(data = DEgenes, aes(x = logFC, y = -log(adj.P.Val), color = ((-log(adj.P.Val) > 3) & (logFC > 1 | logFC < -1))))+
  scale_colour_manual(name = 'BH p-value < 0.05', values = setNames(c('red','black'),c(T, F)), labels = c("False", "True"))+
  geom_point()+
  geom_vline(xintercept=0)+
  geom_vline(xintercept=-1)+
  geom_vline(xintercept=1)+
  geom_hline(yintercept=3)+
  geom_text(aes(label = ifelse((-log(adj.P.Val) > 5) & (logFC > 6 | logFC < -6), gene, "")), vjust=-1, size = 3)+
  #geom_text(aes(label = ifelse((-log(adj.P.Val) > 5) | (logFC > 2.5 | logFC < -2.5), gene, "")), vjust=-1, size = 3)+#(to visualize some of the genes, note overlaps)
  ylab("-log(p-value)")+
  xlab("logFC")+
  labs(title="Gene expression differences in two groups")+
  theme_bw()+
  theme(axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        #panel.border = element_blank(),
        panel.background = element_blank()) +
  theme(axis.title.x=element_text(size=12),
        axis.text.y=element_text(size=12),
        axis.title.y=element_text(size=14),
        axis.ticks.x=element_blank(),
        strip.text.x = element_text(size=14),
        strip.background = element_rect(colour="white", fill="white"),
        legend.text=element_text(size=10),
        legend.title=element_text(size=10))
dev.off()
```
```{r}
############################################################
#Which genes are differentially activated in CAR Tregs (UT vs 19 vs HA) using Treg only?
#A Compare the 3 groups of Tregs (UT, 19, HA; 9 samples) using PCA, heat map (with genes) and venn diagram. (does it make sense?)

#B *DE genes:**Which genes and transcriptional pathways are
# differentially activated in HA CAR Treg vs the others (UT and 19) ?*Tregs
# only: HA vs (UT and 19): Volcano plot and pathway analysis.

group1A <- c("64/ha/26/Treg", "55/ha/24/Treg", "52/ha/24/Treg")
group1B <- c("64/ut/26/Treg", "64/cd19/26/Treg", "52/ut/24/Treg", "52/cd19/24/Treg", "55/ut/24/Treg", "55/cd19/24/Treg") 

DGEres <- performDGEanalysis(x, meta_data, group1A, group1B, termGO = termGO,
                             termTF = termTF, termH = termH)

#table with all genes and their scores
DEgenes <- DGEres[["DEgenesTable"]]
#up regulated genes in group1A compared to group1B
upRegulatedGenes <- DGEres[["upRegulatedGenes"]]
#down regulated genes in group1A compared to group1B
downRegulatedGenes <- DGEres[["downRegulatedGenes"]]
#pathways positively enriched in group1A compared to group1B
sign_pos_pathways <- DGEres[["Sign_Pos_Pathway"]]
#pathways negatively enriched in group1A compared to group1B
sign_neg_pathways <- DGEres[["Sign_Neg_Pathway"]]

#pathways positively enriched in group1A compared to group1B
sign_pos_pathwaysH <- DGEres[["Sign_HALLMARK_Pos_Pathway"]]
#pathways negatively enriched in group1A compared to group1B
sign_neg_pathwaysH <- DGEres[["Sign_HALLMARK_Neg_Pathway"]]

TFgsea <- DGEres[["TF_enrichment"]]
TFscores <- DGEres[["TF_scores"]]

#Transcription factors positively enriched in group1A compared to group1B
sign_pos_TF <- DGEres[["Sign_Pos_TF"]]
#Transcription factors negatively enriched in group1A compared to group1B
sign_neg_TF <- DGEres[["Sign_Neg_TF"]]

#Decision table (potential use - Venn Diagramm; checking what genes are DE in different conditions)
dt <- DGEres[["DecisionTable"]]
#log cpm matrix for heat map construction
log_cpm <- DGEres[["LogCPM"]]

#saving data
write.table(upRegulatedGenes, file="/home/german/Caroline_RNA_seq/Caro RNAseq/DE_genes/upRegulated_genes_TregHA_TregUT19.txt", sep="\t", row.names = F)
write.table(downRegulatedGenes, file="/home/german/Caroline_RNA_seq/Caro RNAseq/DE_genes/downRegulated_genes_TregHA_TregUT19.txt", sep="\t", row.names = F)

write.table(sign_pos_TF[,-8], file="/home/german/Caroline_RNA_seq/Caro RNAseq/TF/posTF_TregHA_TregUT19.txt", sep="\t", row.names = F)
write.table(sign_neg_TF[,-8], file="/home/german/Caroline_RNA_seq/Caro RNAseq/TF/negTF_TregHA_TregUT19.txt", sep="\t", row.names = F)

write.table(sign_pos_pathwaysH[,-8], file="/home/german/Caroline_RNA_seq/Caro RNAseq/PathwaysHallmark/posPathways_TregHA_TregUT19.txt", sep="\t", row.names = F)
write.table(sign_neg_pathwaysH[,-8], file="/home/german/Caroline_RNA_seq/Caro RNAseq/PathwaysHallmark/negPathways_TregHA_TregUT19.txt", sep="\t", row.names = F)

write.table(sign_pos_pathways[,-8], file="/home/german/Caroline_RNA_seq/Caro RNAseq/Pathways/posPathways_TregHA_TregUT19.txt", sep="\t", row.names = F)
write.table(sign_neg_pathways[,-8], file="/home/german/Caroline_RNA_seq/Caro RNAseq/Pathways/negPathways_TregHA_TregUT19.txt", sep="\t", row.names = F)

################################################################################
#heatmap construction
#filtering meta data to groups only
meta_data_f <- meta_data %>% filter(Labels %in% c(group1A, group1B))
#topGenesRegulated <- cleaned_log_cpm_df[common_regulated_genes,]

#correct for batch - in this case it's donor factor
#batch <- x$samples$Donor
#batch <- meta_data_f$Batch
#lcpm.no.batches <- removeBatchEffect(log_cpm, batch) #uncomment if you want to use batches

#log_cpm_f <- lcpm.no.batches #uncomment if you want to use batches
log_cpm_f <- log_cpm
colnames(log_cpm_f) <- meta_data_f$Labels
my_sample_col_an <- my_sample_col[colnames(log_cpm_f),]

upRegulatedGenes <- upRegulatedGenes$gene
downRegulatedGenes <- downRegulatedGenes$gene
common_regulated_genes <- c(upRegulatedGenes[1:50], downRegulatedGenes[1:50])
topGenesRegulated <- log_cpm_f[common_regulated_genes,]
#topGenesRegulated <- topGenesRegulated[,col.order]

#scaling the data (without it in pheatmap specify scale = "row")
topGenesRegulated <- t(scale(t(topGenesRegulated)))
rownames(my_sample_col_an) <- colnames(topGenesRegulated)

#creating a heatmap with pheatmap (no clustering because we ordered samples by time point)
pheatmap(topGenesRegulated, cluster_rows = T, cluster_cols = T, 
         color=bluered(21), scale = "none", clustering_method = "ward.D2", 
         clustering_distance_cols = "euclidean", show_colnames = T, show_rownames = F, 
         annotation_col = my_sample_col_an,
        main = "Clustering heatmap for 50 common regulated genes")
```
```{r}
################################################################################
#Volcano plot construction
devSVG(file = "/home/german/Caroline_RNA_seq/Caro RNAseq/TregHA_TregUT19_volcanoplot.svg")
ggplot(data = DEgenes, aes(x = logFC, y = -log(adj.P.Val), color = ((-log(adj.P.Val) > 3) & (logFC > 1 | logFC < -1))))+
  scale_colour_manual(name = 'BH p-value < 0.05', values = setNames(c('red','black'),c(T, F)), labels = c("False", "True"))+
  geom_point()+
  geom_vline(xintercept=0)+
  geom_vline(xintercept=-1)+
  geom_vline(xintercept=1)+
  geom_hline(yintercept=3)+
  geom_text(aes(label = ifelse((-log(adj.P.Val) > 5) & (logFC > 3 | logFC < -3), gene, "")), vjust=-1, size = 3)+
  #geom_text(aes(label = ifelse((-log(adj.P.Val) > 5) | (logFC > 2.5 | logFC < -2.5), gene, "")), vjust=-1, size = 3)+#(to visualize some of the genes, note overlaps)
  ylab("-log(p-value)")+
  xlab("logFC")+
  labs(title="Gene expression differences in two groups")+
  theme_bw()+
  theme(axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        #panel.border = element_blank(),
        panel.background = element_blank()) +
  theme(axis.title.x=element_text(size=12),
        axis.text.y=element_text(size=12),
        axis.title.y=element_text(size=14),
        axis.ticks.x=element_blank(),
        strip.text.x = element_text(size=14),
        strip.background = element_rect(colour="white", fill="white"),
        legend.text=element_text(size=10),
        legend.title=element_text(size=10))
#dev.off()
```


